File: frontend/src/api/communicationTemplates.ts
/**
 * Communication Templates API Client
 *
 * This module provides a client for managing communication templates,
 * including creating, reading, updating, and deleting
 * template configurations for candidate communications.
 *
 * @example
 * ```ts
 * import { communicationTemplatesClient } from '@/api/communicationTemplates';
 *
 * // List all communication templates
 * const templates = await communicationTemplatesClient.listTemplates();
 *
 * // Create a new template
 * const newTemplate = await communicationTemplatesClient.createTemplate({
 *   name: 'Interview Invitation',
 *   type: 'email',
 *   subject: 'Interview Invitation - {{job_title}}',
 *   body: 'Dear {{candidate_name}}, we would like to invite you...'
 * });
 *
 * // Preview a template with variables
 * const preview = await communicationTemplatesClient.previewTemplate('template-id', {
 *   candidate_name: 'John Doe',
 *   job_title: 'Senior Developer'
 * });
 *
 * // Update a template
 * const updated = await communicationTemplatesClient.updateTemplate('template-id', {
 *   name: 'Updated Template Name'
 * });
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';

/**
 * Communication template type enum
 */
export type TemplateType = 'email' | 'sms' | 'in_system';

/**
 * Template category enum
 */
export type TemplateCategory = 'outreach' | 'interview' | 'rejection' | 'offer' | 'follow_up' | 'other';

/**
 * Template create request
 */
export interface CommunicationTemplateCreate {
  name: string;
  type: TemplateType;
  category?: TemplateCategory;
  subject?: string;
  body: string;
  variables?: string[];
  description?: string;
  is_active?: boolean;
}

/**
 * Template update request
 */
export interface CommunicationTemplateUpdate {
  name?: string;
  type?: TemplateType;
  category?: TemplateCategory;
  subject?: string;
  body?: string;
  variables?: string[];
  description?: string;
  is_active?: boolean;
}

/**
 * Template response
 */
export interface CommunicationTemplateResponse {
  id: string;
  name: string;
  type: TemplateType;
  category: TemplateCategory | null;
  subject: string | null;
  body: string;
  variables: string[];
  description: string | null;
  is_active: boolean;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  usage_count?: number;
}

/**
 * Template list response
 */
export interface CommunicationTemplateListResponse {
  total: number;
  templates: CommunicationTemplateResponse[];
}

/**
 * Template preview request
 */
export interface TemplatePreviewRequest {
  variables: Record<string, string | number | boolean>;
}

/**
 * Template preview response
 */
export interface TemplatePreviewResponse {
  subject?: string;
  body: string;
  rendered_at: string;
}

/**
 * API error response
 */
export interface ApiError {
  detail: string;
  status?: number;
}

/**
 * Default API configuration for communication templates client
 */
const DEFAULT_CONFIG = {
  baseURL: import.meta.env.VITE_API_URL ?? '',
  timeout: 10000, // 10 seconds
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * Communication Templates API Client class
 *
 * Provides methods for managing communication template configurations with proper
 * error handling and type safety.
 */
export class CommunicationTemplatesClient {
  private client: AxiosInstance;

  /**
   * Create a new CommunicationTemplates client instance
   *
   * @param config - Optional configuration overrides
   */
  constructor(config: Partial<typeof DEFAULT_CONFIG> = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };

    this.client = axios.create(finalConfig);

    // Response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(this.transformError(error))
    );
  }

  /**
   * Transform Axios error to standardized API error
   *
   * @param error - Axios error
   * @returns Transformed API error
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string }>;

    // Network error (no response)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Request timeout. Please check your connection and try again.',
          status: 408,
        };
      }
      return {
        detail: 'Network error. Please check your connection and try again.',
        status: 0,
      };
    }

    // Server returned error response
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Use server's error message if available
    if (data?.detail) {
      return { detail: data.detail, status };
    }

    // Default error messages by status code
    const defaultMessages: Record<number, string> = {
      400: 'Invalid request. Please check your input.',
      401: 'Unauthorized. Please log in.',
      403: 'Forbidden. You do not have permission.',
      404: 'Template not found.',
      409: 'A template with this name already exists.',
      422: 'Validation error. Please check your input.',
      429: 'Too many requests. Please try again later.',
      500: 'Server error. Please try again later.',
      502: 'Bad gateway. Please try again later.',
      503: 'Service unavailable. Please try again later.',
    };

    return {
      detail: data?.detail || defaultMessages[status] || 'An unexpected error occurred.',
      status,
    };
  }

  /**
   * Create a communication template
   *
   * @param request - Create request with template details
   * @returns Created template
   * @throws ApiError if creation fails
   *
   * @example
   * ```ts
   * const template = await communicationTemplatesClient.createTemplate({
   *   name: 'Interview Invitation',
   *   type: 'email',
   *   subject: 'Interview Invitation - {{job_title}}',
   *   body: 'Dear {{candidate_name}}, we would like to invite you...',
   *   variables: ['candidate_name', 'job_title', 'interview_date']
   * });
   * ```
   */
  async createTemplate(request: CommunicationTemplateCreate): Promise<CommunicationTemplateResponse> {
    try {
      const response: AxiosResponse<CommunicationTemplateResponse> = await this.client.post(
        '/api/communications/templates/',
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * List communication templates with optional filters
   *
   * @param skip - Number of records to skip (pagination)
   * @param limit - Maximum number of records to return
   * @param type - Optional filter by template type
   * @param category - Optional filter by template category
   * @param search - Optional filter by name (case-insensitive partial match)
   * @param isActive - Optional filter by active status
   * @returns List of communication templates
   * @throws ApiError if listing fails
   *
   * @example
   * ```ts
   * // Get all templates
   * const templates = await communicationTemplatesClient.listTemplates();
   *
   * // Get only email templates
   * const emailTemplates = await communicationTemplatesClient.listTemplates(0, 100, 'email');
   *
   * // Search by name
   * const interviewTemplates = await communicationTemplatesClient.listTemplates(
   *   0,
   *   100,
   *   undefined,
   *   undefined,
   *   'interview'
   * );
   * ```
   */
  async listTemplates(
    skip: number = 0,
    limit: number = 100,
    type?: TemplateType,
    category?: TemplateCategory,
    search?: string,
    isActive?: boolean
  ): Promise<CommunicationTemplateListResponse> {
    try {
      const params: Record<string, number | string | boolean> = { skip, limit };
      if (type) params.type = type;
      if (category) params.category = category;
      if (search) params.search = search;
      if (isActive !== undefined) params.is_active = isActive;

      const response: AxiosResponse<CommunicationTemplateListResponse> = await this.client.get(
        '/api/communications/templates/',
        { params }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get a specific template by ID
   *
   * @param templateId - Template ID
   * @returns Template details
   * @throws ApiError if not found
   *
   * @example
   * ```ts
   * const template = await communicationTemplatesClient.getTemplate('template-uuid');
   * ```
   */
  async getTemplate(templateId: string): Promise<CommunicationTemplateResponse> {
    try {
      const response: AxiosResponse<CommunicationTemplateResponse> = await this.client.get(
        `/api/communications/templates/${templateId}`
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Update a template
   *
   * @param templateId - Template ID
   * @param request - Update request with fields to modify
   * @returns Updated template
   * @throws ApiError if update fails
   *
   * @example
   * ```ts
   * const updated = await communicationTemplatesClient.updateTemplate('template-uuid', {
   *   name: 'Updated Template Name',
   *   subject: 'New Subject Line'
   * });
   * ```
   */
  async updateTemplate(
    templateId: string,
    request: CommunicationTemplateUpdate
  ): Promise<CommunicationTemplateResponse> {
    try {
      const response: AxiosResponse<CommunicationTemplateResponse> = await this.client.put(
        `/api/communications/templates/${templateId}`,
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Delete a template
   *
   * @param templateId - Template ID
   * @throws ApiError if deletion fails
   *
   * @example
   * ```ts
   * await communicationTemplatesClient.deleteTemplate('template-uuid');
   * ```
   */
  async deleteTemplate(templateId: string): Promise<void> {
    try {
      await this.client.delete(`/api/communications/templates/${templateId}`);
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Preview a template with variable substitution
   *
   * @param templateId - Template ID
   * @param request - Preview request with variable values
   * @returns Preview with rendered template
   * @throws ApiError if preview fails
   *
   * @example
   * ```ts
   * const preview = await communicationTemplatesClient.previewTemplate('template-uuid', {
   *   variables: {
   *     candidate_name: 'John Doe',
   *     job_title: 'Senior Developer',
   *     interview_date: '2024-02-15'
   *   }
   * });
   * ```
   */
  async previewTemplate(
    templateId: string,
    request: TemplatePreviewRequest
  ): Promise<TemplatePreviewResponse> {
    try {
      const response: AxiosResponse<TemplatePreviewResponse> = await this.client.post(
        `/api/communications/templates/${templateId}/preview`,
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get the underlying Axios instance
   *
   * This is useful for making custom requests not covered by the convenience methods.
   *
   * @returns Axios instance
   */
  getAxiosInstance(): AxiosInstance {
    return this.client;
  }
}

/**
 * Default communication templates client instance
 *
 * Use this singleton instance for all communication templates calls.
 */
export const communicationTemplatesClient = new CommunicationTemplatesClient();

/**
 * Export communication templates client class for custom instances
 */
export default CommunicationTemplatesClient;

File: backend/api/communication_templates.py
"""
Communication template management endpoints.

This module provides endpoints for managing communication templates,
including CRUD operations for creating, reading, updating, and deleting
templates with variable substitution and preview capabilities.
"""
import logging
import re
from typing import Dict, List, Optional

from fastapi import APIRouter, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

router = APIRouter()


class TemplateEntry(BaseModel):
    """Individual template entry definition."""

    organization_id: str = Field(..., description="ID of the organization")
    name: str = Field(..., description="Name of the template", min_length=1, max_length=255)
    type: str = Field(..., description="Type of communication (email, sms, phone)", min_length=1, max_length=50)
    subject: Optional[str] = Field(None, description="Subject line (for emails)", max_length=500)
    body: str = Field(..., description="Body content of the communication", min_length=1)
    variables: Dict[str, str] = Field(default_factory=dict, description="Template variables with descriptions")
    is_active: bool = Field(True, description="Whether this template is active")
    language: str = Field("en", description="Language code (e.g., 'en', 'ru')", min_length=2, max_length=10)
    created_by: Optional[str] = Field(None, description="ID of the user creating this template")


class TemplateCreate(BaseModel):
    """Request model for creating a template."""

    template: TemplateEntry = Field(..., description="Template to create")


class TemplateUpdate(BaseModel):
    """Request model for updating a template."""

    name: Optional[str] = Field(None, description="Name of the template", min_length=1, max_length=255)
    type: Optional[str] = Field(None, description="Type of communication (email, sms, phone)", min_length=1, max_length=50)
    subject: Optional[str] = Field(None, description="Subject line (for emails)", max_length=500)
    body: Optional[str] = Field(None, description="Body content of the communication", min_length=1)
    variables: Optional[Dict[str, str]] = Field(None, description="Template variables with descriptions")
    is_active: Optional[bool] = Field(None, description="Whether this template is active")
    language: Optional[str] = Field(None, description="Language code (e.g., 'en', 'ru')", min_length=2, max_length=10)


class TemplatePreviewRequest(BaseModel):
    """Request model for template preview."""

    variables: Dict[str, str] = Field(..., description="Variable values for substitution")


class TemplateResponse(BaseModel):
    """Response model for a single template."""

    id: str = Field(..., description="Unique identifier for the template")
    organization_id: str = Field(..., description="ID of the organization")
    name: str = Field(..., description="Name of the template")
    type: str = Field(..., description="Type of communication")
    subject: Optional[str] = Field(None, description="Subject line")
    body: str = Field(..., description="Body content")
    variables: Dict[str, str] = Field(..., description="Template variables")
    is_active: bool = Field(..., description="Whether template is active")
    language: str = Field(..., description="Language code")
    created_by: Optional[str] = Field(None, description="Creator user ID")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")


class TemplateListResponse(BaseModel):
    """Response model for listing templates."""

    templates: List[TemplateResponse] = Field(..., description="List of templates")
    total_count: int = Field(..., description="Total number of templates")


def _extract_template_variables(content: str) -> List[str]:
    """
    Extract template variables from content.

    Variables are expected to be in the format {{variable_name}}.

    Args:
        content: Content to extract variables from

    Returns:
        List of variable names found in the content
    """
    pattern = r"\{\{(\w+)\}\}"
    variables = re.findall(pattern, content)
    return list(set(variables))


def _substitute_variables(content: Optional[str], variables: Dict[str, str]) -> str:
    """
    Substitute template variables in content.

    Args:
        content: Content with variables in {{variable_name}} format
        variables: Dictionary of variable values

    Returns:
        Content with variables substituted
    """
    if not content:
        return ""

    result = content
    for key, value in variables.items():
        result = result.replace(f"{{{{{key}}}}}", str(value))

    return result


@router.post(
    "/",
    response_model=TemplateResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["Communication Templates"],
)
async def create_template(request: TemplateCreate) -> JSONResponse:
    """
    Create a communication template.

    This endpoint creates a new communication template with support for
    variable substitution in subject and body. Templates can be used for
    emails, SMS, and other communication types.

    Args:
        request: Create request with template details

    Returns:
        JSON response with created template

    Raises:
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {
        ...     "template": {
        ...         "organization_id": "org123",
        ...         "name": "Interview Invitation",
        ...         "type": "email",
        ...         "subject": "Interview Invitation - {{company_name}}",
        ...         "body": "Dear {{candidate_name}}, ...",
        ...         "variables": {"candidate_name": "Candidate name", "company_name": "Company name"}
        ...     }
        ... }
        >>> response = requests.post("http://localhost:8000/api/communication-templates/", json=data)
        >>> response.json()
    """
    try:
        logger.info(f"Creating communication template: {request.template.name}")

        # Validate template type
        valid_types = ["email", "sms", "phone"]
        if request.template.type not in valid_types:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=f"Type must be one of: {', '.join(valid_types)}",
            )

        # Validate body content
        if not request.template.body or not request.template.body.strip():
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Body content cannot be empty",
            )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        template_response = {
            "id": "placeholder-id",
            "organization_id": request.template.organization_id,
            "name": request.template.name,
            "type": request.template.type,
            "subject": request.template.subject,
            "body": request.template.body,
            "variables": request.template.variables,
            "is_active": request.template.is_active,
            "language": request.template.language,
            "created_by": request.template.created_by,
            "created_at": "2024-01-25T00:00:00Z",
            "updated_at": "2024-01-25T00:00:00Z",
        }

        logger.info(f"Created template: {template_response['id']}")

        return JSONResponse(
            status_code=status.HTTP_201_CREATED,
            content=template_response,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create template: {str(e)}",
        ) from e


@router.get("/", tags=["Communication Templates"])
async def list_templates(
    organization_id: Optional[str] = Query(None, description="Filter by organization ID"),
    type: Optional[str] = Query(None, description="Filter by communication type"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    language: Optional[str] = Query(None, description="Filter by language code"),
) -> JSONResponse:
    """
    List communication templates with optional filters.

    Args:
        organization_id: Optional organization ID filter
        type: Optional communication type filter
        is_active: Optional active status filter
        language: Optional language code filter

    Returns:
        JSON response with list of templates

    Raises:
        HTTPException(500): If database query fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/communication-templates/?type=email")
        >>> response.json()
    """
    try:
        logger.info(
            f"Listing templates with filters - organization_id: {organization_id}, "
            f"type: {type}, is_active: {is_active}, language: {language}"
        )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        response_data = {
            "templates": [],
            "total_count": 0,
        }

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error listing templates: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list templates: {str(e)}",
        ) from e


@router.get("/{template_id}", tags=["Communication Templates"])
async def get_template(template_id: str) -> JSONResponse:
    """
    Get a specific communication template by ID.

    Args:
        template_id: Unique identifier of the template

    Returns:
        JSON response with template details

    Raises:
        HTTPException(404): If template is not found
        HTTPException(500): If database query fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/communication-templates/123e4567-e89b-12d3-a456-426614174000")
        >>> response.json()
    """
    try:
        logger.info(f"Getting template: {template_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "id": template_id,
                "organization_id": "org123",
                "name": "Interview Invitation",
                "type": "email",
                "subject": "Interview Invitation - {{company_name}}",
                "body": "Dear {{candidate_name}},\n\nWe are pleased to invite you...",
                "variables": {
                    "candidate_name": "Candidate name",
                    "company_name": "Company name",
                    "interview_date": "Interview date"
                },
                "is_active": True,
                "language": "en",
                "created_by": "user123",
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            },
        )

    except Exception as e:
        logger.error(f"Error getting template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get template: {str(e)}",
        ) from e


@router.put("/{template_id}", tags=["Communication Templates"])
async def update_template(
    template_id: str, request: TemplateUpdate
) -> JSONResponse:
    """
    Update a communication template.

    Args:
        template_id: Unique identifier of the template
        request: Update request with fields to modify

    Returns:
        JSON response with updated template

    Raises:
        HTTPException(404): If template is not found
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {"name": "Updated Interview Invitation", "is_active": False}
        >>> response = requests.put(
        ...     "http://localhost:8000/api/communication-templates/123",
        ...     json=data
        ... )
        >>> response.json()
    """
    try:
        logger.info(f"Updating template: {template_id}")

        # Validate template type if provided
        if request.type is not None:
            valid_types = ["email", "sms", "phone"]
            if request.type not in valid_types:
                raise HTTPException(
                    status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                    detail=f"Type must be one of: {', '.join(valid_types)}",
                )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "id": template_id,
                "organization_id": "org123",
                "name": request.name if request.name is not None else "Interview Invitation",
                "type": request.type if request.type is not None else "email",
                "subject": request.subject if request.subject is not None else "Interview Invitation - {{company_name}}",
                "body": request.body if request.body is not None else "Dear {{candidate_name}}...",
                "variables": request.variables if request.variables is not None else {},
                "is_active": request.is_active if request.is_active is not None else True,
                "language": request.language if request.language is not None else "en",
                "created_by": "user123",
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update template: {str(e)}",
        ) from e


@router.delete("/{template_id}", tags=["Communication Templates"])
async def delete_template(template_id: str) -> JSONResponse:
    """
    Delete a communication template.

    Args:
        template_id: Unique identifier of the template

    Returns:
        JSON response confirming deletion

    Raises:
        HTTPException(404): If template is not found
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.delete("http://localhost:8000/api/communication-templates/123")
        >>> response.json()
        {"message": "Template deleted successfully"}
    """
    try:
        logger.info(f"Deleting template: {template_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={"message": f"Template {template_id} deleted successfully"},
        )

    except Exception as e:
        logger.error(f"Error deleting template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete template: {str(e)}",
        ) from e


@router.post("/{template_id}/preview", tags=["Communication Templates"])
async def preview_template(
    template_id: str, request: TemplatePreviewRequest
) -> JSONResponse:
    """
    Preview a communication template with variable substitution.

    This endpoint renders the template with provided variable values,
    allowing users to preview the final content before sending.

    Args:
        template_id: Unique identifier of the template
        request: Preview request with variable values

    Returns:
        JSON response with rendered subject and body

    Raises:
        HTTPException(404): If template is not found
        HTTPException(500): If preview generation fails

    Examples:
        >>> import requests
        >>> data = {
        ...     "variables": {
        ...         "candidate_name": "John Doe",
        ...         "company_name": "Acme Corp",
        ...         "interview_date": "2024-02-01"
        ...     }
        ... }
        >>> response = requests.post(
        ...     "http://localhost:8000/api/communication-templates/123/preview",
        ...     json=data
        ... )
        >>> response.json()
        {
            "subject": "Interview Invitation - Acme Corp",
            "body": "Dear John Doe,\\n\\nWe are pleased to invite you...",
            "missing_variables": []
        }
    """
    try:
        logger.info(f"Previewing template: {template_id}")

        # For now, use placeholder template data
        # Database integration will be added in a later subtask when we have async session setup
        template_subject = "Interview Invitation - {{company_name}}"
        template_body = "Dear {{candidate_name}},\n\nWe are pleased to invite you to interview for the position at {{company_name}} on {{interview_date}}.\n\nBest regards"

        # Extract expected variables from template
        subject_vars = _extract_template_variables(template_subject)
        body_vars = _extract_template_variables(template_body)
        all_vars = set(subject_vars + body_vars)

        # Find missing variables
        provided_vars = set(request.variables.keys())
        missing_vars = list(all_vars - provided_vars)

        # Substitute variables
        rendered_subject = _substitute_variables(template_subject, request.variables)
        rendered_body = _substitute_variables(template_body, request.variables)

        response_data = {
            "subject": rendered_subject,
            "body": rendered_body,
            "missing_variables": missing_vars,
        }

        logger.info(f"Generated preview for template: {template_id}")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error previewing template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to preview template: {str(e)}",
        ) from e
